web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx

```diff
@@ -53,31 +129,95 @@ export const ConnectMyComputerContextProvider: FC<{
       }
   );
 
-  const runAgentAndWaitForNodeToJoin = useCallback(async () => {
-    await connectMyComputerService.runAgent(props.rootClusterUri);
+  const [downloadAgentAttempt, downloadAgent] = useAsync(
+    useCallback(async () => {
+      setCurrentAttempt('download');
+      await connectMyComputerService.downloadAgent();
+    }, [connectMyComputerService])
+  );
+
+  const [startAgentAttempt, startAgent] = useAsync(
+    useCallback(async () => {
+      setCurrentAttempt('start');
+      await connectMyComputerService.runAgent(props.rootClusterUri);
+
+      const abortController = new AbortController();
+      try {
+        const server = await Promise.race([
+          connectMyComputerService.waitForNodeToJoin(
+            props.rootClusterUri,
+            abortController.signal
+          ),
+          throwOnAgentProcessErrors(
+            mainProcessClient,
+            props.rootClusterUri,
+            abortController.signal
+          ),
+          wait(20_000, abortController.signal).then(() => {
+            throw new Error(
+              'The agent did not manage to join the cluster within 20 seconds.'
+            );
+          }),
+        ]);
+        setCurrentAttempt('process');
+        return server;
+      } catch (error) {
+        // in case of any error kill the agent
+        await connectMyComputerService.killAgent(props.rootClusterUri);
+        throw error;
+      } finally {
+        abortController.abort();
+      }
+    }, [connectMyComputerService, mainProcessClient, props.rootClusterUri])
+  );
 
-    // TODO(gzdunek): Replace with waiting for the node to join.
-    const waitForNodeToJoin = wait(1_000);
+  const downloadAndStartAgent = async () => {
+    const [, error] = await downloadAgent();
+    if (error) {
+      return;
+    }
+    await startAgent();
+  };
```

Comment:
I feel like this should be defined in the Status component since it's not reused anywhere else.

Reply:
I assume that the auto-start feature will be kept in the context and will need to call this function, so I think it is better to keep it here?

Reply:
Cool, I forgot about this in the context of auto-start.